src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -2801,6 +2803,86 @@ exprt c_typecheck_baset::do_special_functions(
     overflow.type() = bool_typet{};
     return overflow;
   }
+  else if(
+    identifier == "__builtin_add_overflow" ||
+    identifier == "__builtin_sub_overflow" ||
+    identifier == "__builtin_mul_overflow")
+  {
+    if(expr.arguments().size() != 3)
+    {
+      std::ostringstream error_message;
+      error_message << expr.source_location().as_string() << ": " << identifier
+                    << " takes exactly 3 arguments, but "
+                    << expr.arguments().size() << " were provided";
+      throw invalid_source_file_exceptiont{error_message.str()};
+    }
+
+    for(std::size_t arg_index = 0; arg_index < 3; ++arg_index)
+    {
+      auto const &argument = expr.arguments()[arg_index];
+      bool const arg_has_correct_type =
+        (arg_index < 2) ? (argument.type().id() == ID_signedbv ||
```

Comment:
Trying to do this in a single loop makes this a little complex to read. Would it be better to handle the argument type-check (seeing as their are only 3 args) as two separate blocks, one block to check arg 0 and 1, and one to check arg 2, with each block setting the arg_has_correct_type flag, then the error message block could stay the same (but not be in a loop).

Reply:
Hm, maybe? Could you sketch out how you’d do it?

Reply:
Sorry, yeah, some pseudo code would have been better :-)
```
bool correct_type=true;
// Typecheck args 0 and 1
for(int i = 0; i<1; i++) {
   correct_type &= is_correct_type(arg[i])
}
// Typecheck arg 2
correct_type &= is_correct_result_type(arg[2]);

// Error message if typecheck fails
if(!correct_type) {
 // user error message here
}
```

Reply:
So basically the same logic you have, but without the tertiary operator which you effectively use to encode additional state in the argument index :-)

Reply:
Oh, and I think you may be able to use the existing utility functions `is_signed_or_unsigned_bitvector`, `is_c_integral_type` and  `is_c_integral_pointer_type` here too ?

Reply:
I didn’t know about these, I’ll see about clarifying it using your suggestions.